sub_circuit_identification/src/write_verilog_lef.py

```diff
@@ -108,6 +110,61 @@ def map_pins(self, a, b):
             print("unmatched ports found")
             return 0
 
+class WriteSpice:
+    """ write hierarchical verilog file """
+
+    def __init__(self, circuit_graph, circuit_name, inout_pin_names,subckt_list):
+        self.circuit_graph = circuit_graph
+        self.circuit_name = circuit_name
+        self.inout_pins = inout_pin_names
+        self.pins = inout_pin_names
+        self.subckt_list = subckt_list
+
+    def check_ports_match(self,port,subckt):
+        for members in self.subckt_list:
+            if members["name"]==subckt and port in members["ports"]:
+                return 1
+
+    def print_subckt(self, fp):
+        logging.info("Writing module : %s", self.circuit_name)
+        fp.write("\n.subckt " + self.circuit_name + " ")
+        fp.write(' '.join(self.pins))
+
+        for node, attr in self.circuit_graph.nodes(data=True):
+            if 'source' in attr['inst_type']:
+                logging.info("Skipping source nodes : %s", node)
+                continue
+            if 'net' not in attr['inst_type']:
+                logging.info("Writing node: %s", attr['inst_type'])
+                fp.write("\n" + node + ' '+ attr['real_inst_type'] + ' ')
+                ports = []
+                nets = []
+                if "ports_match" in attr:
+                    logging.info("Nets connected to ports: %s",attr["ports_match"])
+                    for key, value in attr["ports_match"].items():
+                        ports.append(key)
+                        nets.append(value)
+                    if 'DCL_NMOS' in attr['inst_type']:
+                        nets[1:1]=[nets[0]]
+                    elif 'DCL_NMOS' in attr['inst_type']:
+                        nets[1:1]=[nets[2]]
+                else:
+                    logging.error("No connectivity info found : %s",
+                                 ', '.join(attr["ports"]))
+                    ports = attr["ports"]
+                    nets = list(self.circuit_graph.neighbors(node))
+
+                fp.write(' '.join(nets) + self.all_values(attr['values']) )
```

Comment:
Bulk pin connection appears to have been discarded at some point.

Reply:
Yes since our cell generator doesn't take body, these are getting dropped. I will add a body connection based on device type.

Reply:
Thank you ! This should work in the short term. In the long run, we probably should be dealing with bulk connections at the primitive level as well. I know we have discussed this and avoided it multiple times but it may be good to bring up once more even if we just create dummy pins at the primitive level and route it up later.